instana/propagators/base_propagator.py

```diff
@@ -41,85 +40,177 @@ class BasePropagator():
+            if disable_traceparent == "":
+                ctx = SpanContext(span_id=tp_parent_id, trace_id=tp_trace_id[-16:],
+                                  level=int(level.split(",")[0]) if level else 1, baggage={}, sampled=True,
+                                  synthetic=synthetic is not None)
+
+                ctx.trace_parent = True
+
+                if tracestate and "in=" in tracestate:
+                    instana_ancestor = self.__ts.get_instana_ancestor(tracestate)
+                    ctx.instana_ancestor = instana_ancestor
+
+                ctx.long_trace_id = tp_trace_id
+            else:
+                try:
+                    if tracestate and "in=" in tracestate:
+                        instana_ancestor = self.__ts.get_instana_ancestor(tracestate)
+
+                        ctx = SpanContext(span_id=instana_ancestor.p, trace_id=instana_ancestor.t,
+                                          level=int(level.split(",")[0]) if level else 1, baggage={}, sampled=True,
+                                          synthetic=synthetic is not None)
+
+                        ctx.instana_ancestor = instana_ancestor
+                except Exception:
+                    logger.debug("extract instana ancestor error:", exc_info=True)
+
+            self.__set_correlation_properties(level, ctx)
+
+        elif trace_id and span_id:
+
+            ctx = SpanContext(span_id=span_id, trace_id=trace_id, level=int(level.split(",")[0]) if level else 1,
+                              baggage={}, sampled=True, synthetic=synthetic is not None)
+
+        elif synthetic:
+            ctx = SpanContext(synthetic=synthetic)
+
+        ctx.traceparent = traceparent
+        ctx.tracestate = tracestate
+
+        return ctx
+
+    @staticmethod
+    def __set_correlation_properties(level, ctx):
+        try:
+            if level and "correlationType" in level:
+                ctx.correlation_type = level.split(",")[1].split("correlationType=")[1].split(";")[0]
+            if level and "correlationId" in level:
+                ctx.correlation_id = level.split(",")[1].split("correlationId=")[1].split(";")[0]
+        except Exception:
+            logger.debug("extract instana correlation type/id error:", exc_info=True)
+
+    def __extract_instana_headers(self, dc):
+        """
+        Search carrier for the *HEADER* keys and return the tracing key-values
+
+        :param dc: The dict or list potentially containing context
+        :return: trace_id, span_id, level, synthetic, traceparent, tracestate
+        """
+        trace_id, span_id, level, synthetic, traceparent, tracestate = [None] * 6
+
+        # Headers can exist in the standard X-Instana-T/S format or the alternate HTTP_X_INSTANA_T/S style
+        try:
+            trace_id = dc.get(self.LC_HEADER_KEY_T) or dc.get(self.ALT_LC_HEADER_KEY_T) or dc.get(
+                self.LC_HEADER_KEY_T.encode("utf-8")) or dc.get(self.ALT_LC_HEADER_KEY_T.encode("utf-8"))
```

Comment:
because for example the asgi instrumentation uses BINARY format for the header extraction and then the key, is not "x-instana-t" but b'x-instana-t'

Reply:
Do you know why we use BINARY there?

Reply:
ASGI has the headers are bytestrings https://asgi.readthedocs.io/en/latest/specs/www.html, we use it for fastapi and starlette instrumentation. FastApi and Starlette seem to have ways to parse the header into some Header objects not binary, but this does not seem that trivial, it will affect the instrumentations. In general, the extract method was always checking if the headers are coming as binary and decoding them. I am doing the same thing, just changed a bit the way, as there is no need currently to for loop through all incoming headers, decode if needed and then go through a number of if conditions. In my opinion the binary extract is needed, the binary inject though of the 2 W3C headers I will check if it is needed or not. If no HTTP framework used binary inject, I will take the 2 headers out of the inject.